formats/html.ts

```diff
@@ -21,24 +73,340 @@ export class HtmlImporter extends FormatImporter {
+				.filter(({ attachment }) => attachment === "skipped")
+				.map(({ link: { path } }) => path));
+
+			mdContent = transformedAST.map(({ text }) => text).join("");
+			await this.app.vault.modify(mdFile, mdContent);
+		} catch (e) {
+			console.error(e);
+			results.failed.push(file.toString());
+			if (mdFile) {
+				try {
+					await this.app.vault.delete(mdFile);
+				} catch (e) {
+					console.error(e);
+				}
+			}
+		}
+		return results;
+	}
+
+	downloadAttachmentCached(mdFile: TFile, url: URL) {
+		return this.attachments[url.href] ??= this.downloadAttachment(mdFile, url);
+	}
+
+	async downloadAttachment(mdFile: TFile, url: URL) {
+		try {
+			let response;
+			switch (url.protocol) {
+				case "file:":
+					response = await this.requestFile(url);
+					break;
+				case "https:":
+				case "http:":
+					response = await this.requestHTTP(url);
+					break;
+				default:
+					throw new Error(url.href);
+			}
+			if (!await this.filterAttachment(response)) {
+				return "skipped";
+			}
+			let filename = getURLFilename(url);
+			const { data, mime: actualMime } = response;
+			if ((mime(splitFilename(filename).extension) || "application/octet-stream") !== actualMime) {
+				const ext = extension(actualMime);
+				if (ext) {
+					filename += `.${ext}`;
+				}
+			}
+			return await this.writeAttachment(mdFile, filename, data);
+		} catch (e) {
+			console.error(e);
+			return "failed";
+		}
+	}
+
+	async requestFile(url: URL) {
+		const data = (await fsPromises.readFile(nodeUrl.fileURLToPath(url.href))).buffer;
+		return { mime: await detectMime(url, data), data };
+	}
+
+	async requestHTTP(url: URL) {
+		url = new URL(url.href);
+		let response;
+		try {
+			url.protocol = "https:";
+			response = await requestUrl({
+				url: url.href,
+				method: "GET",
+				throw: true,
+			});
+		} catch (e) {
+			try {
+				url.protocol = "http:";
+				response = await requestUrl({
+					url: url.href,
+					method: "GET",
+					throw: true,
+				});
+			} catch (e2) {
+				console.error(e2);
+				throw e;
+			}
+		}
+		const { arrayBuffer: data } = response;
+		return { mime: response.headers["Content-Type"] || await detectMime(url, data), data };
+	}
 
-		for (let file of files) {
+	async writeAttachment(mdFile: TFile, filename: string, data: ArrayBufferLike) {
+		const { basename, extension } = splitFilename(sanitizeFileName(filename));
+		let error;
+		for (let retry = 0; retry < 5; ++retry) {
 			try {
-				let htmlContent = await file.readText();
-				let mdContent = htmlToMarkdown(htmlContent);
-				await this.saveAsMarkdownFile(folder, file.basename, mdContent);
-				results.total++;
+				//@ts-ignore
+				const path: string = await this.app.vault.getAvailablePathForAttachments(basename, extension, mdFile);
+				return await this.app.vault.createBinary(path, data);
 			} catch (e) {
-				console.error(e);
-				results.failed.push(file.toString());
+				// retry in case `path` is the same for multiple invocations of `writeAttachment`
```

Comment:
This can be solved by un-parallelizing the calls.

Reply:
Did that by creating a small utility `PromiseExecutor`.